exercises/custom-set/example.rs

```diff
@@ -1,149 +1,100 @@
 
-    pub fn is_empty(&self) -> bool {
-        self.len() == 0
+    pub fn add(&mut self, element: T) {
+        if !self.contains(&element) {
+            self.collection.push(element)
+        }
     }
 
-    pub fn contains(&self, needle: &T) -> bool {
-        for element in self.elements.iter() {
-            if needle == element {
-                return true;
-            }
-        }
-        false
+    pub fn delete(&mut self, element: T) {
+        self.collection.retain(|x| x != &element)
     }
 
-    pub fn insert(&mut self, element: T) -> bool {
-        if !self.contains(&element) {
-            self.elements.push(element);
-            return true;
-        }
-        false
+    pub fn contains(&self, other: &T) -> bool {
+        self.collection.contains(other)
     }
 
-    pub fn remove(&mut self, needle: &T) -> bool {
-        let mut index: Option<usize> = None;
-        for (i, element) in self.iter().enumerate() {
-            if needle == element {
-                index = Some(i);
-                break;
-            }
-        }
-        match index {
-            Some(index) => {
-                self.elements.remove(index);
-                true
-            }
-            None => false
-        }
+    pub fn is_empty(&self) -> bool {
+        self.collection.is_empty()
     }
-    
-    pub fn difference(&self, other: &CustomSet<T>) -> CustomSet<T> {
-        let mut difference: CustomSet<T> = CustomSet::new();
-        for element in self.elements.iter() {
-            if !other.contains(element) {
-                difference.insert(element.clone());
-            }
-        }
-        difference
+
+    pub fn size(&self) -> usize {
+        self.collection.len()
     }
 
-    pub fn intersection(&self, other: &CustomSet<T>) -> CustomSet<T> {
-        let mut intersection: CustomSet<T> = CustomSet::new();
-        for element in self.elements.iter() {
-            if other.contains(element) {
-                intersection.insert(element.clone());
-            }
-        }
-        intersection
+    pub fn is_subset(&self, other: &Self) -> bool {
+        self.collection.iter().all(|x| other.contains(x))
     }
 
-    pub fn union(&self, other: &CustomSet<T>) -> CustomSet<T> {
-        let mut union = CustomSet::new();
-        for element in self.elements.iter().chain(other.elements.iter()) {
-            union.insert(element.clone());
-        }
-        union
+    pub fn is_disjoint(&self, other: &Self) -> bool {
+        !self.collection.iter().any(|x| other.contains(x))
     }
 
-    pub fn is_disjoint(&self, other: &CustomSet<T>) -> bool {
-        for element in self.elements.iter() {
+    pub fn intersection(&self, other: &Self) -> CustomSet<T> {
+        let mut s = Vec::new();
+        for element in self.collection.iter() {
             if other.contains(element) {
-                return false;
+                s.push(element.clone());
             }
         }
-        true
+        CustomSet::new(s)
```

Comment:
could possibly be written in terms of `filter`, but I don't see this as a hard requirement